File: backend/src/models/project.py
"""
Project data models.

Defines all Pydantic models for projects including:
- API request/response models
- Internal DynamoDB model
"""

from typing import List, Optional
from pydantic import BaseModel, Field, HttpUrl


# API Request Models

class ProjectCreate(BaseModel):
    """Project creation request model."""
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    longDescription: Optional[str] = Field(None, description="Detailed project description")
    tech: List[str] = Field(default_factory=list, description="Technologies used")
    company: Optional[str] = Field(None, max_length=200)
    featured: bool = False
    githubUrl: Optional[HttpUrl] = None
    liveUrl: Optional[HttpUrl] = None
    imageUrl: Optional[HttpUrl] = None


class ProjectUpdate(BaseModel):
    """Project update request model."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, min_length=1)
    longDescription: Optional[str] = None
    tech: Optional[List[str]] = None
    company: Optional[str] = Field(None, max_length=200)
    featured: Optional[bool] = None
    githubUrl: Optional[HttpUrl] = None
    liveUrl: Optional[HttpUrl] = None
    imageUrl: Optional[HttpUrl] = None


# API Response Models

class ProjectResponse(BaseModel):
    """Project API response model."""
    id: str
    name: str
    description: str
    longDescription: Optional[str]
    tech: List[str]
    company: Optional[str]
    featured: bool
    status: str = Field(..., description="Publication status: DRAFT or PUBLISHED")
    githubUrl: Optional[str]
    liveUrl: Optional[str]
    imageUrl: Optional[str]
    createdAt: str = Field(..., description="ISO 8601 timestamp")
    updatedAt: str = Field(..., description="ISO 8601 timestamp")


class ProjectListResponse(BaseModel):
    """Project list response with pagination."""
    items: List[ProjectResponse]
    count: int
    lastEvaluatedKey: Optional[dict] = Field(None, description="Pagination token for next page")

File: frontend/src/types/project.ts
/**
 * Project type definitions.
 *
 * Defines TypeScript interfaces for project data structures
 * matching the backend API schema.
 */

/**
 * Project entity from backend API (uppercase status).
 */
export interface Project {
  id: string;
  name: string;
  description: string;
  longDescription?: string;
  tech: string[];
  company?: string;
  githubUrl?: string;
  liveUrl?: string;
  imageUrl?: string;
  featured: boolean;
  status: 'PUBLISHED' | 'DRAFT';
  createdAt: string;
  updatedAt: string;
  publishedAt?: string;
}

/**
 * Normalized project for frontend use (lowercase status).
 */
export interface ProjectNormalized {
  id: string;
  name: string;
  description: string;
  longDescription?: string;
  tech: string[];
  company?: string;
  githubUrl?: string;
  liveUrl?: string;
  imageUrl?: string;
  featured: boolean;
  status: 'published' | 'draft';
  createdAt: string;
  updatedAt: string;
  publishedAt?: string;
}

/**
 * Data required to create a new project.
 */
export interface ProjectCreate {
  name: string;
  description: string;
  longDescription?: string;
  tech: string[];
  company?: string;
  githubUrl?: string;
  liveUrl?: string;
  imageUrl?: string;
  featured?: boolean;
}

/**
 * Data for updating an existing project.
 */
export interface ProjectUpdate {
  name?: string;
  description?: string;
  longDescription?: string;
  tech?: string[];
  company?: string;
  githubUrl?: string;
  liveUrl?: string;
  imageUrl?: string;
  featured?: boolean;
}

/**
 * Paginated project list response.
 */
export interface ProjectListResponse {
  items: Project[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Query parameters for filtering and pagination.
 */
export interface ProjectListParams {
  status?: 'published' | 'draft' | 'all';
  featured?: boolean;
  limit?: number;
  offset?: number;
}
